csharp/ql/lib/semmle/code/csharp/security/auth/InsecureDirectObjectReferenceQuery.qll

```diff
@@ -0,0 +1,52 @@
+/** Definitions for the Insecure Direct Object Reference query */
+
+import csharp
+import semmle.code.csharp.dataflow.flowsources.Remote
+import ActionMethods
+
+/**
+ * Holds if `m` is a method that may require checks
+ * against the current user to modify a particular resource.
+ */
+// We exclude admin methods as it may be expected that an admin user should be able to modify any resource.
+// Other queries check that there are authorization checks in place for admin methods.
+private predicate needsChecks(ActionMethod m) { m.isEdit() and not m.isAdmin() }
+
+private Expr getParentExpr(Expr ex) { result = ex.getParent() }
+
+/**
+ * Holds if `m` has a parameter or access a remote flow source
+ * that may indicate that it's used as the ID for some resource
+ */
+private predicate hasIdParameter(ActionMethod m) {
+  exists(RemoteFlowSource src | src.getEnclosingCallable() = m |
+    src.asParameter().getName().toLowerCase().matches(["%id", "%idx"])
+    or
+    // handle cases like `Request.QueryString["Id"]`
+    exists(StringLiteral idStr |
+      idStr.getValue().toLowerCase().matches(["%id", "%idx"]) and
+      getParentExpr*(src.asExpr()) = getParentExpr*(idStr)
+    )
+  )
+}
+
+/** Holds if `m` at some point in its call graph may make some kind of check against the current user. */
+private predicate checksUser(ActionMethod m) {
+  exists(Callable c, string name | name = c.getName().toLowerCase() |
+    name.matches(["%user%", "%session%"]) and
+    not name.matches("%get%by%") and // methods like `getUserById` or `getXByUsername` aren't likely to be referring to the current user
+    m.calls*(c)
+  )
+}
+
+/**
+ * Holds if `m` is a method that modifies a particular resource based on
+ * and ID provided by user input, but does not check anything based on the current user
+ * to determine if they should modify this resource.
+ */
+predicate hasInsecureDirectObjectReference(ActionMethod m) {
+  needsChecks(m) and
+  hasIdParameter(m) and
+  not checksUser(m) and
```

Comment:
I am concerned by the rate at which this may produce false positives. My understanding is that e.g. in ASP.NET the standard mechanism for managing authentication/authorisation is [through attributes](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/simple?view=aspnetcore-7.0). The query does not seem to handle those (as well as auth code inherited from parent classes / any other form of auth), so would this query trigger alerts in those cases?

Reply:
Are authorization attributes alone able to manage authorization that's based on the request parameters? 
In [these MRVA results](https://gist.github.com/joefarebrother/a4f8daefefc3910fbf656c2514ebcdc6) I don't see a lot of evidence of other auth mechanisms that were missed.

Reply:
No, using attributes is only sufficient where authorisation is based on roles, policies, etc. 

For [resource-based authorization](https://learn.microsoft.com/en-us/aspnet/core/security/authorization/resourcebased?view=aspnetcore-7.0) checks in the body of the handler are indeed necessary. Would the example in that article be identified by `checksUser`?

However, role-based or policy-based authorisation may be sufficient in a lot of cases, so we probably shouldn't flag up insecure direct object reference vulnerabilities where checks are performed in that way. 

Based on your description of the MRVA experiment, it seems like the results are only for the top 100 repos? Are those the top 100 C# repos or the top 100 repos that use ASP.NET?  In any case, I would probably use a larger sample. It is also likely that the different authorisation mechanisms offered by ASP.NET are more prevalent in enterprise codebases, rather than in open-source ones.

Reply:
> because it uses the `User` property

Ah, I don't think I had realised that properties count as `Callable`s. 

> However in [this top 1000](https://gist.github.com/joefarebrother/2b23e26a640b923e73b35d2e02cebd79) MRVA run, it looks like there are indeed some FPs from role-based authentication for admin roles; so this is something that should be checked.

Thank you for running that experiment! It does indeed seem to confirm my concerns here. 

> Should all instances of the `Authorize` attribute with `Roles` or `Policy` fields set be considered sufficient, or should a more complex check be made?

I think the logic is a little bit more complicated because the `Authorize` attribute may be placed on the method, class, or parent classes(?). So you need to check in all of these places for that attribute as well as any attributes which inherit from `Authorize`. Furthermore, you also need to check for the absence of the `AllowAnonymous` attribute, which may override any `Authorize`-like attribute that's placed further out.

In terms of the arguments to the `Authorize` attribute, I don't think we should care at all. As long as there is some authorisation check, that should be considered a mitigation for the purpose of this query since we can't reason about what the intended audience for a particular endpoint is.

Reply:
It is my understanding that an `Authorize` attribute with no arguments means "Any logged in user can access this", whereas this query focuses on cases where we'd expect stronger access controls than that (e.g. deleting any resource). Is this correct?

Reply:
Yes, although consider that there might be web applications where most visitors have no accounts and the only authorised users you'll get are administrators that should be able to perform such actions. (E.g. a blog, a simple CMS/news website, etc.) In those cases just being logged in may be sufficient access control.

Also, for classes that inherit from `Authorize`, we wouldn't necessarily expect any parameters either, e.g.:

```csharp
class MyAuthorizeAttribute : AuthorizeAttribute
{
    public MyAuthorizeAttribute()
    {
        base.Policy = "Administrators";
    }
}
```